source/MRMesh/MRSerializer.cpp

```diff
@@ -27,15 +29,282 @@
+{
+    std::vector<MeshTexture> result;
+    result.reserve( model.images.size() );
+    for ( auto& image : model.images )
+    {
+        result.emplace_back();
+        auto& meshTexture = result.back();
+        meshTexture.resolution = { image.width, image.height };
+        meshTexture.pixels.resize( image.width * image.height );
+
+        memcpy( meshTexture.pixels.data(), image.image.data(), image.width * image.height * sizeof( Color ) );
+    }
+
+    return result;
+}
+
+std::vector<Material> readMaterials( const tinygltf::Model& model )
+{
+    std::vector<Material> result;
+    result.reserve( model.materials.size() );
+    for ( auto& material : model.materials )
+    {
+        result.emplace_back();
+        auto& curMaterial = result.back();
+
+        auto colorIt = material.values.find( "baseColorFactor" );
+        if ( colorIt != material.values.end() )
+        {
+            const auto& comps = colorIt->second.number_array;
+            curMaterial.baseColor = Color( float( comps[0] ), float( comps[1] ), float( comps[2] ), float( comps[3] ) );
+        }
+        else
+        {
+            curMaterial.baseColor = Color::white();
+        }
+
+        curMaterial.textureIndex = material.pbrMetallicRoughness.baseColorTexture.index;
+    }
+
+    return result;
+}
+
+tl::expected<int, std::string> readVertCoords( VertCoords& vertexCoordinates, const tinygltf::Model& model, const tinygltf::Primitive& primitive )
+{
+    if ( primitive.mode != TINYGLTF_MODE_TRIANGLES )
+        return tl::make_unexpected( "This topology is not implemented" );
+
+    auto posAttrib = primitive.attributes.find( "POSITION" );
+    if ( posAttrib == primitive.attributes.end() )
+        return tl::make_unexpected( "No vertex data" );
+
+    auto accessor = model.accessors[posAttrib->second];
+    auto bufferView = model.bufferViews[accessor.bufferView];
+    auto buffer = model.buffers[bufferView.buffer];
+
+    if ( accessor.componentType != TINYGLTF_COMPONENT_TYPE_FLOAT || accessor.type != TINYGLTF_TYPE_VEC3 )
+        return tl::make_unexpected( "This vertex component type is not implemented" );
+
+    VertId start = VertId( vertexCoordinates.size() );
+    vertexCoordinates.resize( vertexCoordinates.size() + accessor.count );
+    memcpy( &vertexCoordinates[VertId( start )], &buffer.data[accessor.byteOffset + bufferView.byteOffset], accessor.count * sizeof( Vector3f ) );
+
+    return int( accessor.count );
+}
+
+void fillVertsColorMap( Vector<Color, VertId>& vertsColorMap, int vertexCount, const std::vector<Material>& materials, int materialIndex )
+{
+    const auto startPos = vertsColorMap.size();
+    vertsColorMap.resize( vertsColorMap.size() + vertexCount );
+    std::fill( ( uint32_t* )( vertsColorMap.data() + startPos ),
+                ( uint32_t* )( vertsColorMap.data() + startPos + vertexCount ),
+                materialIndex >= 0 ? materials[materialIndex].baseColor.getUInt32() : 0xFFFFFFFF );
```

Comment:
do we call this if there is no per verts colors? (if we are it is better to just set front faces color)